qubes/api/admin.py

```diff
@@ -1630,6 +1631,74 @@ async def vm_device_set_required(self, endpoint, untrusted_payload):
         await self.dest.devices[devclass].update_assignment(dev, mode)
         self.app.save()
 
+    @qubes.api.method(
+        "admin.vm.device.denied.List",
+        no_payload=True, scope="local", read=True)
+    async def vm_device_denied_list(self):
+        """
+        List all denied device interfaces for the VM.
+
+        Returns a newline-separated string.
+        """
+        self.enforce(not self.arg)
+
+        self.fire_event_for_permission()
+
+        denied = self.dest.devices_denied
+        return "\n".join(map(repr, DeviceInterface.from_str_bulk(denied)))
+
+    @qubes.api.method(
+        "admin.vm.device.denied.Add", scope="local", write=True)
+    async def vm_device_denied_add(self, untrusted_payload):
+        """
+        Add device interface(s) to the denied list for the VM.
+
+        Payload:
+            Encoded device interface (can be repeated without any separator).
```

Comment:
What is intended behavior on duplicates? IIUC right now if at least one interface you try to add is already on deny list, the whole operation will fail. Is that intentional? IMHO better would be to ignore duplicates (as in - adding interface that is already denied is no-op, but others non-duplicated added at the same time are still added). Excluding duplicates at the client side is racy (somebody can modify the value between client reads it (to exclude duplicates) and writes it back).

Reply:
Yes, this was intentional; I followed the "fail fast" principle. However, now I understand that in the mentioned example, admin methods should be idempotent. 
To keep consistency, in the case of removing an interface that is not currently on the list nothing should be done, am I right?

Reply:
Yes, I think so